### ISSUE TITLE
[BUG] Random coredump in custom all-reduce with SGLang + large CUDA graph + long prompt (Qwen3.5-397B-FP8)

### ISSUE THREAD

**Issue by zovonoir** [ROCm/aiter#3515]
## Summary

A customer reported random coredumps when serving **Qwen3.5-397B-A17B-FP8** with **TP=4** on **SGLang + ATOM plugin + CUDA graph enabled + long prompt** on AMD MI308X. After extensive debugging we have root-caused the crash to a missing exit barrier in aiter's custom all-reduce kernel `cross_device_reduce_1stage`. A fix is proposed in **#3514**.

This issue tracks the bug itself (symptoms / reproduction / evidence). The proposed fix and review discussion live in the PR.

### Customer-visible

- After some decode steps, the streaming response ends prematurely
- Subsequent requests get `Connection refused` (server is dead)
- Reproduces intermittently, more frequently with long prompts and sampling (`temperature > 0`)

### Server-side

```
:0:rocdevice.cpp ... HSA_STATUS_ERROR_EXCEPTION:
  An HSAIL operation resulted in a hardware exception. code: 0x1016
Fatal Python error: Aborted
Kernel Name: _ZN2at6native25_assert_async_cuda_kernelIbEEvPKT_NS0_3MsgE
[SIGQUIT] Subprocess scheduler_X crashed with exit code -6
```

All 4 TP scheduler subprocesses abort synchronously.

### Conditions that trigger the crash

| Condition | Required? |
|---|---|
| SGLang launcher (`python3 -m sglang.launch_server`, monolithic CUDA-graph capture) | yes |
| CUDA graph enabled (`--disable-cuda-graph` avoids it) | yes |
| aiter custom AR active during graph capture (`registered_input=True` path) | yes |
| Long prompt (≥ ~3K tokens), so decode has enough steps to hit the race | yes |
| `temperature > 0` (sampling) for the crash to manifest | yes |
| `temperature = 0` (greedy) | **no crash, but silently produces garbage tokens** (see below) |

### Greedy vs sampling

Under our diagnostic probe (see below), the underlying AR data race fires at the same ~1% rate in both greedy and sampling modes. The difference is only downstream:

- `argmax(NaN logits)` does not raise → greedy keeps decoding garbage tokens, no crash
- `multinomial(softmax(NaN logits))` trips a CUDA-side probability-validity assert → HSA exception → process abort

So `temperature = 0` is **not** a safe workaround — it just hides the bug.

## Root cause (short version)

`cross_device_reduce_1stage` is the only AR kernel in `csrc/include/custom_all_reduce.cuh` that does not call `end_sync` before kernel exit. This is a **cross-rank write-after-read race**:

1. Fast rank A finishes its peer IPC reads and exits the kernel
2. Rank A's next kernel in the captured CUDA graph runs immediately; PyTorch's caching allocator has already mapped rank A's input slot to that next kernel's output (`graph_pool` does aggressive short-lived slot reuse)
3. Slow rank B is still inside the AR kernel, reading rank A's input slot via IPC
4. Rank B reads whatever rank A's next kernel just wrote there — usually a GEMM intermediate, sometimes bit patterns interpreted as `NaN` / `Inf`
5. Rank B's AR output is therefore wrong; the bad value propagates downstream and (with sampling) eventually trips a CUDA assert

Every other AR kernel variant in the file (`1stage_naive`, `2stage`, `2stage_naive`, the `2stage_write_mode` variants) already calls `end_sync` before exit. `cross_device_reduce_1stage` is the only one missing it.

### Environment

- Image: `rocm/pytorch:rocm7.2.2_ubuntu22.04_py3.10_pytorch_release_2.9.1`
- Hardware: AMD MI308X × 4 (XGMI fully connected)
- Model: Qwen3.5-397B-A17B-FP8, **real weights** (random / dummy weights do NOT exercise the AR call pattern that triggers the bug)
- TP=4
- Framework versions:
  - aiter: `cfca70a33` (current `main`)
  - SGLang: `3b62604ce` (upstream)
  - ATOM: latest (any recent version reproduces)
- **ROCm: 7.2.2 (also reproduces on 7.2.3 — version-independent)**

### Step 1 — launch SGLang server with ATOM plugin

```bash
# inside the image above, with the model mounted at $MODEL_PATH
export PYTHONPATH=/path/to/ATOM:/path/to/sglang/python:/path/to/aiter
export HIP_VISIBLE_DEVICES=0,1,2,3
export CUDA_VISIBLE_DEVICES=0,1,2,3
export SGLANG_USE_AITER=1
export SGLANG_USE_AITER_NEW_CA=false
export SGLANG_AITER_FP8_PREFILL_ATTN=0
export SGLANG_EXTERNAL_MODEL_PACKAGE=atom.plugin.sglang.models   # load ATOM as SGLang model plugin

python3 -m sglang.launch_server \
    --served-model-name qwen397 \
    --model-path $MODEL_PATH \
    --cuda-graph-bs 1 2 4 \
    --mem-fraction-static=0.8 \
    --max-running-requests=4 \
    --tp-size=4 \
    --attention-backend=aiter \
    --trust-remote-code \
    --host 0.0.0.0 --port 30000
```

Key flags that matter for the bug:
- `--tp-size=4` — IPC across 4 peers exposes the cross-rank race
- default CUDA graph **on** (no `--disable-cuda-graph`) — required for `graph_pool` slot reuse
- `SGLANG_USE_AITER=1` + `SGLANG_USE_AITER_NEW_CA=false` — selects aiter's classic custom AR (the kernel in question)
- `SGLANG_EXTERNAL_MODEL_PACKAGE=atom.plugin.sglang.models` — loads ATOM model code; ATOM's model defines many AR call sites and (per its plugin contract) defers compilation/inductor to SGLang. SGLang captures the model as it executes in eager, so the captured graph keeps all the short-lived intermediate tensors that allocator then aggressively recycles into IPC-registered AR-input slots.

### Step 2 — trigger one long-prompt sampling request

```python
import requests

prompt = (
    "在大规模分布式深度学习推理服务中，KV cache 的内存管理是性能与稳定性的核心问题。"
    "When the model generates tokens autoregressively, each layer produces a key and value "
    "tensor that must be cached to avoid recomputation. ... "
) * 120   # ~50K characters, ~25K tokens

body = {
    "model": "qwen397",
    "messages": [
        {"role": "system", "content": prompt},
        {"role": "user",   "content": "请用 5 句话总结上述材料的核心观点。"},
    ],
    "stream": True,
    "max_tokens": 256,
    "temperature": 0.7,
}
requests.post("http://127.0.0.1:30000/v1/chat/completions", json=body, stream=True, timeout=600)
```

Within seconds the server aborts with `HSA_STATUS_ERROR_EXCEPTION 0x1016` and all 4 TP scheduler subprocesses exit with code -6.

Repeating the request with varied user queries (same long system prompt) reproduces the crash within a handful of rounds even when the first round happens to survive — this is the pattern used by our reproducer `repro_crash.py` (long system prompt + rotating queries + `temperature=0.7` + `stream=True`).

### Step 3 — confirm symptoms in the server log

```
:0:rocdevice.cpp ... HSA_STATUS_ERROR_EXCEPTION:
  An HSAIL operation resulted in a hardware exception. code: 0x1016
Fatal Python error: Aborted
Kernel Name: _ZN2at6native25_assert_async_cuda_kernelIbEEvPKT_NS0_3MsgE
[SIGQUIT] Subprocess scheduler_X crashed with exit code -6
```

### Notes on greedy (`temperature=0`) reproduction

Greedy decode with a long prompt also reliably triggers the underlying AR data race (see deterministic probe below — ~1.24% AR sums are wrong even in greedy mode), but does **not** produce an HSA crash, because `argmax(NaN)` does not raise. Instead the server keeps streaming garbage tokens. Use sampling (`temperature > 0`) to get a clean hard-crash signal.

### Direct evidence of AR misread (deterministic probe)

To rule out any datapath / numerical hypothesis, we instrumented `Qwen3NextSparseMoeBlock.forward` to replace the AR input with `torch.zeros_like(...) + 1.0` so the AR sum must deterministically equal `4.0` across TP=4 — any other value is unambiguous bug evidence.

| Configuration | AR calls | Mismatches | Observed values when wrong |
|---|---|---|---|
| Baseline (no fix), greedy | 1048 | **13 (1.24%)** | `NaN`, `Inf`, `872`, `1168`, `−2.66×10³⁴`, ... |
| ATOM POSTPONE workaround on, greedy | 1048 | **0** | (all `4.0`) |

POSTPONE moves the AR input from a graph_pool short-lived slot to a long-lived `hidden_states` slot, eliminating the IPC slot-reuse race — directly confirming the root-cause hypothesis. The ATOM-side implementation of this workaround is in [ROCm/ATOM@631ac66](https://github.com/ROCm/ATOM/commit/631ac666a9e65a34ae14abd31f856354fd2270bd) (branch-only, not merged).

### Workarounds (all explained by the same root cause)

These have all been verified to avoid the crash, by making the AR input not land on a graph_pool short-lived slot that gets IPC-registered:

| Workaround | Mechanism |
|---|---|
| `SGLANG_USE_AITER_AR=0` (SGLang native path) | Dispatch picks pynccl, no IPC |
| `--disable-cuda-graph` | All eager, no graph_pool |
| `ATOM_POSTPONE_MOE_ALL_REDUCE=1` ([ROCm/ATOM@631ac66](https://github.com/ROCm/ATOM/commit/631ac666a9e65a34ae14abd31f856354fd2270bd)) | AR input moved to long-lived `hidden_states` slot |
| Force `registered_input=False` in aiter | hipMemcpyAsync to fixed `_pool[\"input\"]` buffer |
| ATOM standalone (inductor piecewise capture) | Each segment has its own pool + inductor fuses ops and removes a lot of transient tensor materialization, so AR-input slots no longer land on the high-reuse short-lived class |

These workarounds are no longer necessary with the fix in #3514.

## Proposed fix

**#3514** — adds a single `end_sync<ngpus, true>(sg, self_sg, rank);` before kernel exit, matching every other AR kernel variant in the same file.

PR includes:
- N=2 controlled experiment (`13/1048` mismatches → `0/2068`, one-sided Fisher's exact p ≈ `6.7×10⁻⁷`, ≈ 4.8 σ)
- Production-path end-to-end test (HSA crash → 0 crashes + full 256-token coherent response)
- `__builtin_trap()` confirmation that `cross_device_reduce_1stage` is the kernel actually dispatched on this path

## Why this hasn't been caught before

`cross_device_reduce_1stage` is exercised heavily by vLLM and aiter standalone, but those use **piecewise** CUDA-graph capture, which limits per-graph AR call density and graph_pool slot reuse — the race window is too narrow to trigger consistently.

SGLang's **monolithic** capture (entire forward in one graph) combined with a large MoE model (~48 layers × multiple AR call sites per layer) is the first deployment we are aware of that hits the race reliably.

There is also a second, independent factor: **inductor / torch.compile changes the captured graph topology**. ATOM standalone runs the model through inductor, which fuses pointwise ops and removes a lot of the intermediate `torch.empty` materializations that would otherwise become short-lived `graph_pool` slots — so even the AR-input tensors that *do* land in `graph_pool` end up on a much less aggressively reused class of slots. SGLang's CUDA-graph runner, by contrast, captures the model as it executes in **eager**, with all the intermediate tensors materialized. When ATOM is loaded as a SGLang plugin, ATOM deliberately disables its own compilation / inductor pass to avoid conflicting with SGLang's runner — so SGLang+ATOM captures an eager-shaped graph too.

This explains the per-configuration crash pattern we observed:

| Configuration | Capture style | AR-input slot class | Crash |
|---|---|---|---|
| SGLang standalone + aiter custom AR | monolithic, eager-shaped | high-reuse short-lived | yes |
| SGLang + ATOM plugin + aiter custom AR | monolithic, eager-shaped (ATOM defers to SGLang) | high-reuse short-lived | yes |
| ATOM standalone + aiter custom AR | piecewise + inductor-fused | low-reuse / many transient tensors fused away | no |

## Suggestion (in addition to the fix)

A multi-GPU unit test that exercises `cross_device_reduce_1stage` with an adversarial allocator pattern (input slots aggressively reused across AR calls within the same captured graph) would catch this class of bug for any AR kernel in the future. Happy to help draft if useful.

**Comment by zovonoir** [comment]
I printed all reduce output in greedy mode (to avoid crash), the correct value should be 4.0 or -9999.0, but printed log shows sporadic abnormal value. 

<img width="979" height="874" alt="Image" src="https://github.com/user-attachments/assets/23f6964c-8cd0-417d-9a04-fd6c21397b63" />
